lib/network/modules/components/nodes/util/ShapeBase.js

```diff
@@ -43,16 +43,23 @@ class ShapeBase extends NodeBase {
    * @param {boolean} selected
    * @param {boolean} hover
    * @param {ArrowOptions} values
+   * @param {function} customRenderer - a custom shape renderer similar to getShape(shape) functions
    * @private
    */
-  _drawShape(ctx, shape, sizeMultiplier, x, y, selected, hover, values) {
+  _drawShape(ctx, shape, sizeMultiplier, x, y, selected, hover, values, customRenderer) {
     this.resize(ctx, selected, hover, values);
     this.left = x - this.width / 2;
     this.top = y - this.height / 2;
 
     this.initContextForDraw(ctx, values);
-    getShape(shape)(ctx, x, y, values.size);
-    this.performFill(ctx, values);
+    if (shape === 'custom') {
+      ctx.save();
+      customRenderer({ ctx, x, y, selected, hover, values });
```

Comment:
One more thing: I would flatten the arguments here (`{ ctx, x, y, selected, hover, ...values }`). Values is very vague and from API point of view, I don't see any reason for its existence. It would be just confusing to users that something can be directly destructured and something is hidden under values. Also someone could change the values and mess with something down the line as it's passed by reference (I'm not sure if this is the case at the moment but at some point it could be, better safe than sorry).

Alternatively it would make sense to group it in some logical way, e.g.:
```typescript
interface CustomShapeArgs {
  ctx: CanvasRenderingContext2D;
  position: { x: number; y: number };
  state: { selected: boolean; hovered: boolean };
  style: { opacity: number; color: string; ... };
}
```

Reply:
ok, makes sense.